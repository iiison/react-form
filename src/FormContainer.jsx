import React, { Component } from 'react'
import PropTypes            from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies

import validations          from './utils/validationRules'

/**
 * Select rule from 3 diffrent places(predefined rules, form level rules or input level rules)
 */
export function selectRule({ ruleValue, customFormRules, customRules }) {
  return validations[ruleValue] || customRules[ruleValue] || customFormRules[ruleValue]
}

/**
 * Validates if the the selected rule(returned from `selectRule`) is valid or not.
 * if rule is valid then tests the input field value agains the selected rule.
 * returns empty string if rule is valid and saticefies the condition.
 */
export function validateRule({ id, value, displayName, ruleValue, ruleArgs, validation }) {
  if (validation) {
    const stringifiedValue = value.toString()
    const isRuleSatisfied = ( ruleValue !== 'required' && !stringifiedValue )
      ? true
      : validation.rule.apply(null, ruleArgs).test(value.toString())
    let error = ''

    if (!isRuleSatisfied) {
      error = validation.formatter.apply(null, [displayName || id, ...ruleArgs])

      return error
    }
  } else {
    throw `invalid validation rule: ${ruleValue}, please use an existing validation rule name or pass a custom function with same name through 'customRules' prop in Input: ${id}. Rule value should be an object with keys: 'rule' as an Regex and 'formatter' as a function, that formats the value.` // eslint-disable-line
  }

  return ''
}


/**
 * Main Component, a HOC, will store all fields' state.
 */
export default class FormContainer extends Component {
  constructor(props) {
    super(props)

    const {
      isDisabled = false,
      shouldValidateForm,
      defaultClasses
    } = this.props

    this.state = {
      shouldValidateForm,
      defaultClasses,
      isDisabled,
      errors     : {},
      isFetching : false
    }
  }

  getChildContext() {
    return {
      setFieldValue : ({ event, field, value, isMultipleValues, id = 'value' }, callBack) => {
        const fieldName = field.id
        const setState = (name, fieldValue) => {
          this.setState((prevState) => ({
            ...prevState,
            fields : {
              ...prevState.fields,
              [fieldName] : {
                ...prevState.fields[name],
                shouldValidateField : true,
                [id]                : fieldValue
              }
            }
          }), callBack)
        }

        if (!value && value !== '') {
          setState(fieldName, event.currentTarget.value)
        } else if (isMultipleValues) {
          const stateCopy = { ...this.state }

          for (const item in value) {
            stateCopy.fields[item].shouldValidateField = true
            stateCopy.fields[item][id] = value[item]
          }

          this.setState((prevState) => ({
            ...prevState,
            ...stateCopy
          }))
        } else {
          setState(fieldName, value)
        }

        return this.state
      },

      addField : (data) => {
        this.setState((prevState) => ({
          ...prevState,
          errors : {
            ...prevState.errors,
            [data.id] : ''
          },
          fields : {
            ...prevState.fields,
            [data.id] : {
              ...data,
              // shouldValidateField : false // Not sure why this was added
            }
          }
        }))
      },

      setFormData : (data) => {
        this.setState((prevState) => ({
          ...prevState,
          ...data
        }))
      },

      validateForm : (fieldName) => {
        const { shouldValidateForm, fields } = this.state
        const field = fields[fieldName]

        if (shouldValidateForm) {
          if (field) {
            if ( field.shouldValidateField ) {
              this.validateField(field)
            }

            return
          }

          for (const unit in fields) {
            const fieldData = fields[unit]

            if (fieldData) {
              this.validateField(fieldData)
            }
          }
        }

        return
      },
      formData : this.state
    }
  }

  /**
   * React Lifecycle Method: Renders the data
   *
   * @return {DOM} Main container DOM.
   */
  render() {
    const { children } = this.props

    return (
      <form action=''>
        {children}
      </form>
    )
  }

  validateField = ({
    id,
    value,
    validate,
    displayName,
    customRules = {}
  }) => {
    let error = ''
    const rules = validate ? validate.split('|') : ''
    const { customValidators : customFormRules } = this.props

    if (rules.length) {
      for (const rule in rules) {
        const ruleName = rules[rule]
        const ruleDetails = ruleName.split('-')
        const [ruleValue, ...ruleArgs] = ruleDetails
        const validation = selectRule({ ruleValue, customFormRules, customRules })

        try {
          error = validateRule({ id, value, displayName, ruleValue, ruleArgs, validation })

          if (error !== '') {
            break
          }
        } catch (invalidRuleError) {
          throw invalidRuleError
        } 
      }

      this.setState((prevState) => ({
        ...prevState,
        errors : {
          ...prevState.errors,
          [id] : error
        }
      }))
    }
  }

  static defaultProps = {
    errors             : {}, // eslint-disable-line
    children           : <div />,
    isDisabled         : false,
    shouldValidateForm : true,
    customValidators   : {},
    defaultClasses     : {
      contClass  : '',
      fieldClass : '',
      errorClass : '',
      labelClass : ''
    }
  }

  static propTypes = {
    children           : PropTypes.node,
    isDisabled         : PropTypes.bool,
    shouldValidateForm : PropTypes.bool,
    customValidators   : PropTypes.object, // eslint-disable-line
    defaultClasses     : PropTypes.shape({
      labelClass : PropTypes.string,
      contClass  : PropTypes.string,
      errorClass : PropTypes.string,
      fieldClass : PropTypes.string,
    })
  }

  static childContextTypes = {
    addField      : PropTypes.func,
    setFieldValue : PropTypes.func,
    validateForm  : PropTypes.func,
    formData      : PropTypes.object, // eslint-disable-line
    setFormData   : PropTypes.func
  }
}

