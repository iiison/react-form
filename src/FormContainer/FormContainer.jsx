import React, { Component } from 'react'

import { Form } from '$COMPONENTS'
import validations from '$UTILS/validationRules'

/**
 * Test all validation rules ofa field
 *
 * @param {Object} fieldData all field realated data
 *
 * @return {String}         validation error ina field
 */
export function testFieldRules(fieldData) {
  const { validate, value, displayName } = fieldData

  const validationRules = validate.split('|')

  for (const validation in validationRules) {
    const { rule, formatter } = validations[validationRules[validation]]()

    const result = rule.test(value)

    if (result === false) {
      return formatter(displayName)
    }
  }
}

/**
 * Generic Form Handler, Handles form state,
 * Pass Form elemnt's JSON throught the FormContainer and
 * it will redner the form.
 */
export default class FormContainer extends Component {
  /**
   * Constructor for the container
   *
   * @param {Object} props form data
   */
  constructor(props) {
    super(props)
    const { formData : { fields } } = props

    this.state = {
      isTested : false,
    }

    this.state.fields = fields.reduce((prev, curr) => {
      const temp = {
        value          : '',
        error          : '',
        shouldBeTested : false,
        ...curr
      }

      return {
        ...prev,
        [temp.id] : temp
      }
    }, {})
  }

  /**
   * DOM Event : Chaanges `shouldBeTested`, will be executed on
   * keyUp in an input field
   *
   * @param {String} field field name
   */
  makeFieldTestable = (field) => {
    const fieldRef = this.state.fields[field]

    fieldRef.shouldBeTested = true

    this.setState((previousState) => {
      return {
        fields : {
          ...previousState.fields,
          [field] : fieldRef
        }
      }
    })
  }

  /**
   * Handle Key press event on the input,
   * tests field value if field is already has error.
   * @param  {String} field Field name in which event was started
   */
  handleKeyPress = (field) => {
    const fieldData = this.state.fields[field]
    const { error } = fieldData

    this.makeFieldTestable(field)

    if (error) {
      this.validateInput(field)
    }
  }

  /**
   * DOM Event: Handles input value change
   *
   * @param {Object} event DOM event object
   */
  handleChange = (event) => {
    event.preventDefault()

    const elemRef = event.target
    const idName = elemRef.id

    this.setState((previousState) => {
      const { fields } = previousState

      return {
        fields : {
          ...fields,
          [idName] : {
            ...fields[idName],
            value : elemRef.value
          }
        }
      }
    })
  }

  /**
   * Validate single input
   *
   * @param  {String}  field       field name to be tested
   * @param  {Boolean} isFinalTest will be true if whole form is being tested
   * @return {String}              validation error of the field
   */
  validateInput = (field, isFinalTest) => {
    const fieldData = this.state.fields[field]
    const { validate } = fieldData

    if ((fieldData.shouldBeTested === true || isFinalTest) && validate) {
      const error = testFieldRules(fieldData) || ''

      this.setState((previousState) => {
        const { fields } = previousState
        const updatedField = {
          ...fields[field],
          error
        }

        return {
          fields : {
            ...fields,
            [field] : updatedField
          }
        }
      })

      return error
    }
  }

  /**
   * Handles Form Submission, callback actual form submit fxn
   * after the form validation. validation is the part of plugin.
   *
   * @param {Object} event DOM Event object
   */
  handleFormSubmit = (event) => {
    event.preventDefault()

    const { handleSubmit } = this.props
    const { fields } = this.state
    const errors = []

    for ( const field in fields ) {
      const error = this.validateInput(field, true)

      error && errors.push(error)
    }

    if (!errors.length) {
      handleSubmit(fields)
    }
  }

  /**
   * React Lifecycle Method: Renders the data
   *
   * @return {DOM} Main container DOM.
   */
  render() {
    const { fields } = this.state
    const {
      error,
      messages,
      isFetching,
      submitButtonText,
      submitButtonClassName } = this.props

    const props = {
      fields,
      messages,
      events : {
        onSubmit          : this.handleFormSubmit,
        onValueChange     : this.handleChange,
        validateInput     : this.validateInput,
        handleKeyPress    : this.handleKeyPress,
        makeFieldTestable : this.makeFieldTestable
      },
      submitButtonData : {
        error,
        isFetching,
        submitButtonText,
        submitButtonClassName,
        onSubmit : this.handleFormSubmit
      }
    }

    return (
      <Form
        {...props}
      />
    )
  }

  /**
  * React Lifecycle Method: Executes right before the
  * component mounts
  */
  componentWillMount() {
  }
}
