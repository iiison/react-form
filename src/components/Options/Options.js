import React, { Component } from 'react'
import PropTypes from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies

export default class Option extends Component {
  getChecboxNewValue = ({ event, field, isChecked }) => {
    const previousValue = field.value || []
    const fieldValue = event.currentTarget.value
    const value = isChecked === true ? [...previousValue, fieldValue] : (() => {
      const previousValueCopy = [...previousValue]
      const index = previousValueCopy.indexOf(fieldValue)

      previousValueCopy.splice(index, 1)

      return previousValueCopy
    })()

    return value
  }

  handleInputChange = ({ event }) => {
    const { id } = this.props
    const { formData : { fields, errors }, setFieldValue, validateForm } = this.context
    const field = fields[id]
    const { options, events, type } = field
    const fieldValue = event.currentTarget.value
    const isChecked = event.currentTarget.checked
    const value = type === 'checkbox'
      ? this.getChecboxNewValue({ event, field, isChecked })
      : fieldValue

    
    setFieldValue({ event, field, value }, () => {
      validateForm(id)
      
      if (field.events.onChange && typeof field.events.onChange === 'function') {
        const { formData } = this.context

        field.events.onChange({
          formData,
          setFieldValue
        })
      }
    })
  }

  drawOptions({ optionClass, optionContClass, optionLabelClass }) {
    const { options, id : fieldID, type } = this.props
    const { formData : { fields } } = this.context
    const field = fields[fieldID]
    const renderedOptions = options.map((option) => {
      const { id, value, displayName } = option

      return (
        <div key={id}>
          <input
            id={id}
            value={value}
            name={fieldID} 
            type={type}
            defaultChecked={field.value.indexOf(value) !== -1}
            onChange={(evt) => {
              const event = { ...evt }

              evt.persist()
              this.handleInputChange({ event })
            }}
          />
          <label htmlFor={id}>{displayName}</label>
        </div>
      )
    })

    return renderedOptions
  }

  render() {
    const {
      id,
      label,
      classes,
      shouldUseDefaultClasses
    } = this.props
    const {
      contClass,
      labelClass,
      errorClass,
      optionClass,
      optionContClass,
      optionLabelClass
    } = classes
    const {
      formData = {}
    } = this.context
    const {
      defaultClasses,
      errors : allErrors
    } = formData
    const {
      contClass  : defaultContClass,
      errorClass : defaultErrorClass,
      labelClass : defaultLabelClass
    } = defaultClasses
    const errors = allErrors[id]

    return (
      <div className={`input-cont ${shouldUseDefaultClasses && defaultContClass} ${contClass}`}>
        {
          label
            ? <div className={`col-12 ${labelClass} ${shouldUseDefaultClasses && defaultLabelClass} label`}>{label}</div>
            : ''
        }
        {(formData.fields && formData.fields[id]) ? this.drawOptions({ optionClass, optionContClass, optionLabelClass }) : ''}
        {
          errors
            ? <div className={`col-12 error ${errorClass} ${shouldUseDefaultClasses && defaultErrorClass}`}>{errors}</div>
            : ''
        }
      </div>
    )
  }

  componentDidMount() {
    this.context.addField(this.props)
  }

  static defaultProps = {
    value                   : [],
    label                   : '',
    events                  : {},
    validate                : '',
    shouldValidateField     : true,
    shouldUseDefaultClasses : true,
    classes                 : {
      contClass           : '',
      labelClass          : '',
      fieldClass          : '',
      errorClass          : '',
      optionClass         : '',
      selectedOptionClass : ''
    }
  }

  static propTypes = {
    id                      : PropTypes.string.isRequired,
    events                  : PropTypes.object,
    classes                 : PropTypes.object,
    label                   : PropTypes.string,
    validate                : PropTypes.string,
    shouldValidateField     : PropTypes.bool,
    shouldUseDefaultClasses : PropTypes.bool,
    displayName             : PropTypes.string.isRequired,
    options                 : PropTypes.array.isRequired,
    type                    : PropTypes.oneOf(['checkbox', 'radio']).isRequired,
    value                   : PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.string
    ]).isRequired
  }

  static contextTypes       = {
    addField      : PropTypes.func.isRequired,
    setFieldValue : PropTypes.func.isRequired,
    validateForm  : PropTypes.func.isRequired,
    formData      : PropTypes.object.isRequired
  }
}

