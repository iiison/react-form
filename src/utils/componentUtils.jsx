import React from 'react'

export function drawElements ({ shouldUseDefaultClasses, classes, content, defaultClasses }) {
  return content
    ? <div className={`col-12 ${classes} ${shouldUseDefaultClasses && defaultClasses}`}>{content}</div>
    : ''
}

