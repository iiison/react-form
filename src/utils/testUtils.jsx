import React, { Component } from 'react'

export function wrapWithContext(context, contextTypes, Elements) {
  class WrapWithContext extends Component {
    getChildContext = () => context

    render() {
      return (
        <div>
          {this.props.children}
        </div>
      ) 
    }

    getChildContext() {
      return context
    }

    static childContextTypes = contextTypes
  }

  return WrapWithContext
}

