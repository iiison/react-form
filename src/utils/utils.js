/**
 * Debounce input function
 *
 * @param   {Function} func function to be debounced
 * @param   {Number}   time time to wait
 *
 * @returns {Function}      debounced function
 */
export function debounce(func, time) {
  let timeout

  return function(...args) {
    const ref = this

    clearTimeout(timeout)
    timeout = setTimeout(() => func.apply(ref, args), time)
  }
}

