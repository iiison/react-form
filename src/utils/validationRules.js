const validations = {
  email() {
    return {
      rule    : /^([\w-]+(?:\.[\w-]+)*)@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$/i,
      formatter (fieldName) {
        const message = '%c is not valid'

        return message.replace('%c', fieldName)
      }
    }
  },
  required() {
    return {
      rule    : /\S/,
      formatter (fieldName) {
        const message = '%c is required'

        return message.replace('%c', fieldName)
      }
    }
  }
}

export default validations
