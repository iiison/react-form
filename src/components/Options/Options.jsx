import React, { Component } from 'react'
import PropTypes from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies
import { drawElements } from '$UTILS/componentUtils'

export default class Option extends Component {
  getChecboxNewValue = ({ event, field, isChecked }) => {
    const previousValue = field.value || []
    const fieldValue = event.currentTarget.value
    const value = isChecked === true ? [...previousValue, fieldValue] : (() => {
      const previousValueCopy = [...previousValue]
      const index = previousValueCopy.indexOf(fieldValue)

      previousValueCopy.splice(index, 1)

      return previousValueCopy
    })()

    return value
  }

  handleInputChange = ({ event }) => {
    const { id } = this.props
    const { formData : { fields }, setFieldValue, validateForm } = this.context
    const field = fields[id]
    const { type } = field
    const fieldValue = event.currentTarget.value
    const isChecked = event.currentTarget.checked
    const value = type === 'checkbox'
      ? this.getChecboxNewValue({ event, field, isChecked })
      : fieldValue

    
    setFieldValue({ event, field, value }, () => {
      validateForm(id)
      
      if (field.events.onChange && typeof field.events.onChange === 'function') {
        const { formData } = this.context

        field.events.onChange({
          formData,
          setFieldValue
        })
      }
    })
  }

  drawOptions({ optionClass, optionContClass, optionLabelClass }) {
    const { options, id : fieldID, type } = this.props
    const { formData : { fields } } = this.context
    const field = fields[fieldID]
    const fieldValueAsList = Array.isArray(field.value) ? field.value : [field.value]
    const renderedOptions = options.map((option) => {
      const { id, value, displayName } = option

      return (
        <div key={id} className={optionContClass}>
          <input
            id={id}
            value={value}
            name={fieldID} 
            type={type}
            className={optionClass}
            defaultChecked={fieldValueAsList.includes(value)}
            onChange={(evt) => {
              const event = { ...evt }

              evt.persist()
              this.handleInputChange({ event })
            }}
          />
          <label htmlFor={id} className={optionLabelClass}>{displayName}</label>
        </div>
      )
    })

    return renderedOptions
  }

  render() {
    const {
      id,
      label,
      classes,
      shouldUseDefaultClasses
    } = this.props
    const {
      contClass,
      labelClass,
      errorClass,
      optionClass,
      optionContClass,
      optionLabelClass
    } = classes
    const {
      formData = {}
    } = this.context
    const {
      defaultClasses,
      errors : allErrors
    } = formData
    const {
      contClass  : defaultContClass,
      errorClass : defaultErrorClass,
      labelClass : defaultLabelClass
    } = defaultClasses
    const errors = allErrors[id]

    return (
      <div className={`input-cont ${shouldUseDefaultClasses && defaultContClass} ${contClass}`}>
        {
          drawElements({
            shouldUseDefaultClasses,
            defaultClasses : defaultLabelClass,
            classes        : `${labelClass} label`,
            content        : label
          })
        }
        {(formData.fields && formData.fields[id]) ? this.drawOptions({ optionClass, optionContClass, optionLabelClass }) : ''}
        {
          drawElements({
            shouldUseDefaultClasses,
            defaultClasses : defaultErrorClass,
            classes        : errorClass,
            content        : errors
          })
        }
      </div>
    )
  }

  componentDidMount() {
    this.context.addField(this.props)
  }

  static defaultProps = {
    displayName             : '',
    label                   : '',
    events                  : {},
    validate                : '',
    shouldValidateField     : true,
    shouldUseDefaultClasses : true,
    classes                 : {
      contClass           : '',
      labelClass          : '',
      fieldClass          : '',
      errorClass          : '',
      optionClass         : '',
      selectedOptionClass : ''
    }
  }

  static propTypes = {
    id                      : PropTypes.string.isRequired,
    label                   : PropTypes.string,
    validate                : PropTypes.string,
    shouldValidateField     : PropTypes.bool,
    shouldUseDefaultClasses : PropTypes.bool,
    displayName             : PropTypes.string,
    options                 : PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
    type                    : PropTypes.oneOf(['checkbox', 'radio']).isRequired,
    events                  : PropTypes.shape({
      onChange : PropTypes.func
    }),
    classes : PropTypes.shape({
      contClass           : PropTypes.string,
      labelClass          : PropTypes.string,
      fieldClass          : PropTypes.string,
      errorClass          : PropTypes.string,
      optionClass         : PropTypes.string,
      selectedOptionClass : PropTypes.string
    }),
    value : PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.string
    ]).isRequired
  }

  static contextTypes = {
    addField      : PropTypes.func.isRequired,
    setFieldValue : PropTypes.func.isRequired,
    validateForm  : PropTypes.func.isRequired,
    formData      : PropTypes.object.isRequired // eslint-disable-line react/forbid-prop-types
  }
}

