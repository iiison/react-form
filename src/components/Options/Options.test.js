import React               from 'react'
import renderer            from 'react-test-renderer'
import PropTypes           from 'prop-types'
import { mount }           from 'enzyme'

import Options               from './Options'
import FormContainer       from '../../FormContainer'
import { wrapWithContext } from '../../utils/testUtils'

/* eslint-disable react/jsx-filename-extension */
const defaultContext = {
  addField      : jest.fn(),
  setFieldValue : jest.fn(),
  validateForm  : jest.fn(),
  formData      : {
    defaultClasses : {
      contClass  : 'default-contClass',
      fieldClass : 'default-fieldClass',
      errorClass : 'default-errorClass',
      labelClass : 'default-labelClass'
    },
    fields : {
      optionsBox : {
        value   : '',
        options : []
      }
    }
  }
}
const contextTypes = {
  addField      : PropTypes.func,
  setFieldValue : PropTypes.func,
  formData      : PropTypes.object,
  validateForm  : PropTypes.func,
}
const customClasses = {
  contClass           : 'options-cont',
  labelClass          : 'options-label',
  fieldClass          : 'options-field',
  errorClass          : 'options-error',
  optionClass         : 'option-unit',
  selectedOptionClass : 'selected-option'
}
const options = [
  {
    id          : 'apple',
    value       : 'apple',
    displayName : 'Fruit Apple'
  },
  {
    id          : 'orange',
    value       : 'orange',
    displayName : 'Fruit orange'
  },
  {
    id          : 'pineapple',
    value       : 'pineapple',
    displayName : 'Fruit Pineapple'
  }
]
const defaultArgs = {
  id                      : 'optionsBox',
  type                    : 'radio',
  value                   : '',
  options                 : [],
  validate                : 'required',
  displayName             : 'Fruits',
  shouldUseDefaultClasses : true
}

describe('>>> Options Container -- Snapshot Test', () => {
  it('Matches the snapshot for radio button with default props', () => {
    const optionsProps = {
      ...defaultArgs,
      options
    }
    const tree = renderer.create(
      <FormContainer>
        <Options {...optionsProps} />
      </FormContainer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot for radio button with custom props', () => {
    const props = {
      ...defaultArgs,
      options,
      label               : 'Options Fruits',
      value               : 'pineapple',
      displayName         : 'Fruits',
      shouldValidateField : true,
      classes             : customClasses
    }

    const tree = renderer.create(
      <FormContainer>
        <Options {...props} />
      </FormContainer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot with error in radio button', () => {
    const customContext = { ...defaultContext }
    const props = {
      ...defaultArgs,
      options,
      onFieldChange : jest.fn(),
      validate      : 'required'
    }

    customContext.formData.fields.optionsBox = props
    customContext.formData.errors = {
      optionsBox : 'some error'
    }

    const WrappedWithContextComponent = wrapWithContext(customContext, contextTypes)
    const tree = renderer.create(
      <WrappedWithContextComponent>
        <Options {...props} />
      </WrappedWithContextComponent>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot for checkbox with default props', () => {
    const optionsProps = {
      ...defaultArgs,
      type : 'checkbox',
      options
    }
    const tree = renderer.create(
      <FormContainer>
        <Options {...optionsProps} />
      </FormContainer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot for checkbox with custom props', () => {
    const props = {
      ...defaultArgs,
      options,
      type                : 'checkbox',
      label               : 'Options Fruits',
      value               : ['pineapple', 'orange'],
      displayName         : 'Fruits',
      shouldValidateField : true,
      classes             : customClasses
    }

    const tree = renderer.create(
      <FormContainer>
        <Options {...props} />
      </FormContainer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Matches the snapshot with error in checkbox', () => {
    const customContext = { ...defaultContext }
    const props = {
      ...defaultArgs,
      options,
      type          : 'checkbox',
      onFieldChange : jest.fn(),
      validate      : 'required'
    }

    customContext.formData.fields.optionsBox = props
    customContext.formData.errors = {
      optionsBox : 'some error'
    }

    const WrappedWithContextComponent = wrapWithContext(customContext, contextTypes)
    const tree = renderer.create(
      <WrappedWithContextComponent>
        <Options {...props} />
      </WrappedWithContextComponent>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
// *************************************************************

// Test DOM Behaviour
describe('>>> Select Box Container -- Shallow Rendering', () => {
  let mountedScreen
  const optionsScreen = (customProps = {}) => {
    const initialState = {
      id    : 'optionsBox',
      value : '',
      type  : 'radio',
      options
    }

    if (!mountedScreen) {
      const updatedState = {
        ...initialState,
        ...customProps
      }

      mountedScreen = mount(<FormContainer><Options {...updatedState} /></FormContainer>)
    }

    return mountedScreen
  }

  beforeEach(() => {
    mountedScreen = undefined
  })

  it('Triggers the change function in radio button', () => {
    const props = {
      events : {
        onChange : jest.fn()
      }
    }

    const tree = optionsScreen(props)
    const optionsInTree = tree.find('input')
    const optionOrange = optionsInTree.at(1)
    const optionPinaple = optionsInTree.at(2)

    optionOrange.simulate('change', {
      currentTarget : {
        checked : true,
        value   : 'orange'
      }
    })

    let selectedInput = tree.find({ defaultChecked : true })

    expect(selectedInput.length).toEqual(1)
    expect(selectedInput.prop('value')).toEqual('orange')
    expect(props.events.onChange).toHaveBeenCalled()

    optionPinaple.simulate('change', {
      currentTarget : {
        checked : true,
        value   : 'pineapple'
      }
    })

    selectedInput = tree.find({ defaultChecked : true })

    expect(selectedInput.length).toEqual(1)

    expect(tree.find('#pineapple').prop('defaultChecked')).toEqual(true)
    expect(tree.find('#orange').prop('defaultChecked')).toEqual(false)
  })

  it('Triggers the change function in checkbox', () => {
    const props = {
      type   : 'checkbox',
      events : {
        onChange : jest.fn()
      }
    }

    const tree = optionsScreen(props)
    const optionsInTree = tree.find('input')
    const optionOrange = optionsInTree.at(1)
    const optionPinaple = optionsInTree.at(2)

    optionOrange.props().onChange({
      persist       : jest.fn(),
      currentTarget : {
        checked : true,
        value   : 'orange'
      }
    })
    tree.update()
    
    let selectedInput = tree.find({ defaultChecked : true })

    expect(selectedInput.length).toEqual(1)
    expect(selectedInput.prop('value')).toEqual('orange')
    expect(props.events.onChange).toHaveBeenCalled()

    optionPinaple.props().onChange({
      persist       : jest.fn(),
      currentTarget : {
        checked : true,
        value   : 'pineapple'
      }
    })
    tree.update()

    selectedInput = tree.find({ defaultChecked : true })

    expect(selectedInput.length).toEqual(2)

    expect(tree.find('#pineapple').prop('defaultChecked')).toEqual(true)
    expect(tree.find('#orange').prop('defaultChecked')).toEqual(true)
    expect(props.events.onChange).toHaveBeenCalled()
  })
})
// *************************************************************
