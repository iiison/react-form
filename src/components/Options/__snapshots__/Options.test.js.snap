// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Options Container -- Snapshot Test Matches the snapshot for checkbox with custom props 1`] = `
<form
  action=""
>
  <div
    className="input-cont  options-cont"
  >
    <div
      className="col-12 options-label label "
    >
      Options Fruits
    </div>
    <div>
      <input
        className="option-unit"
        defaultChecked={false}
        id="apple"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="apple"
      />
      <label
        htmlFor="apple"
      >
        Fruit Apple
      </label>
    </div>
    <div>
      <input
        className="option-unit"
        defaultChecked={true}
        id="orange"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="orange"
      />
      <label
        htmlFor="orange"
      >
        Fruit orange
      </label>
    </div>
    <div>
      <input
        className="option-unit"
        defaultChecked={true}
        id="pineapple"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="pineapple"
      />
      <label
        htmlFor="pineapple"
      >
        Fruit Pineapple
      </label>
    </div>
    
  </div>
</form>
`;

exports[`>>> Options Container -- Snapshot Test Matches the snapshot for checkbox with default props 1`] = `
<form
  action=""
>
  <div
    className="input-cont  "
  >
    
    <div>
      <input
        className=""
        defaultChecked={false}
        id="apple"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="apple"
      />
      <label
        htmlFor="apple"
      >
        Fruit Apple
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="orange"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="orange"
      />
      <label
        htmlFor="orange"
      >
        Fruit orange
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="pineapple"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="pineapple"
      />
      <label
        htmlFor="pineapple"
      >
        Fruit Pineapple
      </label>
    </div>
    
  </div>
</form>
`;

exports[`>>> Options Container -- Snapshot Test Matches the snapshot for radio button with custom props 1`] = `
<form
  action=""
>
  <div
    className="input-cont  options-cont"
  >
    <div
      className="col-12 options-label label "
    >
      Options Fruits
    </div>
    <div>
      <input
        className="option-unit"
        defaultChecked={false}
        id="apple"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="apple"
      />
      <label
        htmlFor="apple"
      >
        Fruit Apple
      </label>
    </div>
    <div>
      <input
        className="option-unit"
        defaultChecked={false}
        id="orange"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="orange"
      />
      <label
        htmlFor="orange"
      >
        Fruit orange
      </label>
    </div>
    <div>
      <input
        className="option-unit"
        defaultChecked={true}
        id="pineapple"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="pineapple"
      />
      <label
        htmlFor="pineapple"
      >
        Fruit Pineapple
      </label>
    </div>
    
  </div>
</form>
`;

exports[`>>> Options Container -- Snapshot Test Matches the snapshot for radio button with default props 1`] = `
<form
  action=""
>
  <div
    className="input-cont  "
  >
    
    <div>
      <input
        className=""
        defaultChecked={false}
        id="apple"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="apple"
      />
      <label
        htmlFor="apple"
      >
        Fruit Apple
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="orange"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="orange"
      />
      <label
        htmlFor="orange"
      >
        Fruit orange
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="pineapple"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="pineapple"
      />
      <label
        htmlFor="pineapple"
      >
        Fruit Pineapple
      </label>
    </div>
    
  </div>
</form>
`;

exports[`>>> Options Container -- Snapshot Test Matches the snapshot with error in checkbox 1`] = `
<div>
  <div
    className="input-cont default-contClass "
  >
    
    <div>
      <input
        className=""
        defaultChecked={false}
        id="apple"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="apple"
      />
      <label
        htmlFor="apple"
      >
        Fruit Apple
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="orange"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="orange"
      />
      <label
        htmlFor="orange"
      >
        Fruit orange
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="pineapple"
        name="optionsBox"
        onChange={[Function]}
        type="checkbox"
        value="pineapple"
      />
      <label
        htmlFor="pineapple"
      >
        Fruit Pineapple
      </label>
    </div>
    <div
      className="col-12  default-errorClass"
    >
      some error
    </div>
  </div>
</div>
`;

exports[`>>> Options Container -- Snapshot Test Matches the snapshot with error in radio button 1`] = `
<div>
  <div
    className="input-cont default-contClass "
  >
    
    <div>
      <input
        className=""
        defaultChecked={false}
        id="apple"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="apple"
      />
      <label
        htmlFor="apple"
      >
        Fruit Apple
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="orange"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="orange"
      />
      <label
        htmlFor="orange"
      >
        Fruit orange
      </label>
    </div>
    <div>
      <input
        className=""
        defaultChecked={false}
        id="pineapple"
        name="optionsBox"
        onChange={[Function]}
        type="radio"
        value="pineapple"
      />
      <label
        htmlFor="pineapple"
      >
        Fruit Pineapple
      </label>
    </div>
    <div
      className="col-12  default-errorClass"
    >
      some error
    </div>
  </div>
</div>
`;
